### DISCUSSION TITLE
Realtime with custom token AND custom users table (auth0 integration)

### DISCUSSION THREAD

**Discussion by benjaminpreiss** [supabase/supabase#13620] [Questions] [answered]
Hey, all of you!

We are using custom tokens that we get from auth0 to authenticate in supabase. That means, we don't really use supabase's internal authentication (the internal users list is empty), but instead just check with the RLS policies if the user_id given in the custom token matches the referenced user_id in the table row. We also have a users table, that is updated from auth0 on every sign in to auth0.

Now we are trying to get realtime to work with this setup, but not really sure how :(

This is how we create our supabase client:

```js
import {
  createClient as createSupabaseClient
} from '@supabase/supabase-js';
const accessToken = $session?.idTokenPayload?.supabase_access_token;
if (accessToken) {
  const options = {
    global: accessToken ?
      {
        headers: {
          Authorization: "Bearer " + accessToken
        }
      } :
      undefined,
    realtime: {
      headers: {
        apikey: accessToken // custom token with 'authenticated' role signed with jwt secret
      },
      params: {
        apikey: PUBLIC_SUPABASE_ANON // supabase anon key
      }
    }
  };
  return createSupabaseClient(PUBLIC_SUPABASE_URL, PUBLIC_SUPABASE_ANON, options);
}
```
Where accessToken is the token we get from auth0 that is signed with the supabase secret key.
this is how the token is created in the auth0 action:

```js
const iat = Math.floor(Date.now() / 1000)
const exp = iat + 60 * 60
const payload = {
  userId: user_id,
  exp: exp,
  iat: iat
}
const tokenSupabase = jwt.sign(payload, event.secrets.SUPABASE_JWT_SECRET)
```

So far, we got our RLS policies working to do all sorts of things except for realtime.

This is how we are trying to `subscribe()` to updates happening to entries that we should have access to:

```js
supabase
  .channel('any')
  .on(
    'postgres_changes', {
      event: 'UPDATE',
      schema: 'public',
      table: 'milestones'
    },
    (payload) => {
      console.log('something changed ', payload);
    }
  )
  .subscribe((status) => console.log(status));
(async () => {
  let {
    data: milestones,
    error: e1
  } = await supabase.from('milestones').select('*');
  if (e1) console.error(e1);
  if (milestones) {
    const {
      data: update_milestones,
      error: e2
    } = await supabase
      .from('milestones')
      .update({
        plan: 'random' + Math.random().toString()
      })
      .eq('id', milestones[0].id);
    console.log(milestones);
    console.log(update_milestones);
    if (e2) console.error(e2);
  }
})();
```

This code works and yields the following `console.log()`:

<img width="1122" alt="Screenshot 2023-04-12 at 11 17 29" src="https://user-images.githubusercontent.com/28349122/231339712-3b06329c-d9d2-444e-a67c-2a597d48db89.png">

As you can also see, it should actually be subscribed... But upon updating one of the entries, we don't get the realtime subscription to log anything...

We have replication enabled for ALL 12 tables in the projects database tab:

<img width="1724" alt="Screenshot 2023-04-12 at 11 27 20" src="https://user-images.githubusercontent.com/28349122/231341131-bd2baa9b-8e74-4f89-8ea1-244979764ae1.png">

The table we are trying to enable realtime for, has the following RLS policies:

<img width="1415" alt="Screenshot 2023-04-12 at 11 28 48" src="https://user-images.githubusercontent.com/28349122/231341343-291486f2-26d4-4c63-b9a4-c958fbc5a25f.png">

... And in detail:

<img width="1163" alt="Screenshot 2023-04-12 at 11 29 43" src="https://user-images.githubusercontent.com/28349122/231341476-9725d7fb-0cf7-4d41-8d71-df880769c57d.png">

<img width="1163" alt="Screenshot 2023-04-12 at 11 29 23" src="https://user-images.githubusercontent.com/28349122/231341424-f086cc52-6231-4670-bce8-86b3f61d4db5.png">

The `auth.is_milestone_owner` function sql script:

```sql
create or replace function auth.is_milestone_owner(owner text, id bigint)
returns boolean
language plpgsql
security definer
set search_path = public
stable
as $$
declare
  is_matched boolean;
begin
  is_matched := false;
  if exists ( 
    select ms.id as id_ms, o.user_id as user_id_o, r.user_id as user_id_r from milestones ms
    inner join matches m on ms.mentoring = m.id
    inner join match_offers o on o.id = m.offer
    inner join match_requests r on r.id = m.request
    where (r.user_id = $1 or o.user_id = $1) and ms.id = $2
  ) then
    is_matched := true;
  end if;
  return is_matched;
end;
$$;
```

And the `auth.user_id` function sql script:

```sql
create or replace function auth.user_id() returns text as $$
  select nullif(current_setting('request.jwt.claims', true)::json->>'userId', '')::text;
$$ language sql stable;
```

But the policies should actually be alright I guess... So what am I doing wrong? Help is greatly appreciated

**Comment by benjaminpreiss** [comment]
We also looked into https://github.com/supabase/supabase-js/issues/553 but couldn't find a solution to our problem there.

**Comment by benjaminpreiss** [comment]
... When initializing the supabase client as referred in the docs (https://supabase.com/docs/guides/realtime/extensions/postgres-changes#custom-tokens)

```js
import {
  createClient as createSupabaseClient
} from '@supabase/supabase-js';
const accessToken = $session?.idTokenPayload?.supabase_access_token;
if (accessToken) {
  const options = {
    global: accessToken ?
      {
        headers: {
          Authorization: "Bearer " + accessToken
        }
      } :
      undefined,
    realtime: {
      headers: {
        apikey: PUBLIC_SUPABASE_ANON
      },
      params: {
        apikey: accessToken
      }
    }
  };
  return createSupabaseClient(PUBLIC_SUPABASE_URL, PUBLIC_SUPABASE_ANON, options);
}
```

... we get a subscription error:

<img width="647" alt="Screenshot 2023-04-12 at 11 59 44" src="https://user-images.githubusercontent.com/28349122/231345910-a1037d28-210d-421d-9166-4c65df65202c.png">

**Reply by rbkayz** [reply]
hey @benjaminpreiss if you manage to figure this out, let me know as well please? Breaking my head over this

**Reply by benjaminpreiss** [reply]
Hahaha, damn! Will ping you :)

I was wondering whether the sql commands in the beginning of [this page](https://supabase.com/docs/guides/realtime/extensions/postgres-changes) might offer a solution... Did you try anything in that direction @rbkayz ?

**Comment by rbkayz** [comment]
We're facing this exact issue as well

we even tried to use the supabse.realtime.setAuth(accessToken), but it still doesn't work. 

We even tried to disable RLS for the tables. The moment we set an access token, there are no events passed (works fine if we leave out the access token)

**Comment by benjaminpreiss** [comment]
I just found that the realtime section of my policies has no RLS policies enabled... I wonder whether the problem might lie here?

<img width="1404" alt="Screenshot 2023-04-12 at 15 46 27" src="https://user-images.githubusercontent.com/28349122/231388777-4a8d80c8-61ef-4e4d-b140-b8d6f1bfa6b2.png">

**Comment by rbkayz** [comment]
hey @benjaminpreiss , figured the issue out

For some reason, the realtime server does not like it if the iat, and exp has decimals. It should be an integer. Please make sure the iat and exp are integers

Additionally, make sure that your custom JWTs include the following keys in the payload:

aud, role, sub

Working fine for me now

**Reply by benjaminpreiss** [reply]
Ahhh, good to know @rbkayz ! Can you maybe let me know what the contents of the fields `aud`, `role`, `sub` are in your case?

**Reply by rbkayz** [reply]
Depends on your rls policies. but typically, aud and role are **authenticated**, and sub is the user id, as defined in the auth.users table

**Reply by benjaminpreiss** [reply]
Ah, damn - I am not sure how to fill these fields, actually. We have used [this guide](https://supabase.com/docs/guides/integrations/auth0#step-7-sign-auth0-token-for-supabase) for setting up the custom token. But we are not really using the `authenticated` role anywhere, since all our RLS policies are based on auth0's user ids

**Comment by benjaminpreiss** [comment]
@GaryAustin1 Do you maybe have an idea on how to solve this problem?

**Reply by GaryAustin1** [reply]
Maybe look at this thread...
https://discord.com/channels/839993398554656828/1075638266893709404/1077018502076051626

**Reply by benjaminpreiss** [reply, marked as answer]
@GaryAustin1 you are a genius! Thank you so much 🥇 Got it to work!
Here is what I had to adjust to whom it may concern:

### Adjust the custom token signed in supabase:
Updated code:
```js
const iat = Math.floor(Date.now() / 1000)
const exp = iat + 60 * 60
const payload = {
  sub: user_id,
  role: "authenticated",
  aud: "authenticated",
  userId: user_id,
  exp: exp,
  iat: iat
}
const tokenSupabase = jwt.sign(payload, event.secrets.SUPABASE_JWT_SECRET)
```
You NEED the properties `sub`, `role` and `aud` (as @rbkayz said), otherwise you'll get a `CHANNEL ERROR`. Fill `role` and `aud` with the value `"authenticated"`, even though you might not really use the authenticated role in your RLS policies.

### Assign `channel.socket.accessToken`

Updated code:
```js
const channel = supabase.channel('any');
channel.socket.accessToken = supabase_access_token;
channel
  .on(
    'postgres_changes', {
      event: 'UPDATE',
      schema: 'public',
      table: 'milestones'
    },
    (payload) => {
      console.log('something changed ', payload);
    }
  )
  .subscribe((status) => console.log(status));
(async () => {
  let {
    data: milestones,
    error: e1
  } = await supabase.from('milestones').select('*');
  if (e1) console.error(e1);
  if (milestones) {
    const {
      data: update_milestones,
      error: e2
    } = await supabase
      .from('milestones')
      .update({
        plan: 'random' + Math.random().toString()
      })
      .eq('id', milestones[0].id);
    console.log(milestones);
    console.log(update_milestones);
    if (e2) console.error(e2);
  }
})();
```

If you don't do `channel.socket.accessToken = supabase_access_token;`, you'll end up getting a `SUBSCRIBED` status but wont get notified on postgres changes.

### Init the client AS DESCRIBED IN THE DOCS

Updated code:
```js
import { createClient as createSupabaseClient } from '@supabase/supabase-js';
const accessToken = supabase_access_token;
if (accessToken) {
  const options = {
    global: {
      headers: {
        Authorization: `Bearer ${accessToken}`
      }
    },
    realtime: {
      headers: {
        apikey: accessToken // custom token with 'authenticated' role signed with jwt secret
      },
      params: {
        apikey: PUBLIC_SUPABASE_ANON // supabase anon key
      }
    }
  };
  return createSupabaseClient(PUBLIC_SUPABASE_URL, PUBLIC_SUPABASE_ANON, options);
```
